src/proj.rs

```diff
@@ -0,0 +1,791 @@
+
+    #[inline(always)]
+    fn warp_p(&self, x: f32, y: f32) -> (f32, f32) {
+        let t = &self.transform;
+        let d = t[6]*x + t[7]*y + t[8];
+
+        ((t[0]*x + t[1]*y + t[2])/d,
+         (t[3]*x + t[4]*y + t[5])/d)
+    }
+
+    #[inline(always)]
+    fn warp_a(&self, x: f32, y: f32) -> (f32, f32) {
+        let t = &self.transform;
+
+        ((t[0]*x + t[1]*y + t[2]),
+         (t[3]*x + t[4]*y + t[5]))
+    }
+
+    #[inline(always)]
+    fn warp_t(&self, x: f32, y: f32) -> (f32, f32) {
+        let t = &self.transform;
+        let tx = t[2];
+        let ty = t[5];
+
+        (x + tx,
+         y + ty)
+    }
+}
+
+/// Combines two projective transformations
+/// (matrix multiplication)
+impl Mul<Proj> for Proj {
+    type Output = Proj;
+
+    fn mul(self, rhs: Proj) -> Proj {
+        use TransformationClass as TC;
+        let t = mul3x3(self.transform, rhs.transform);
+        let i = mul3x3(rhs.inverse, self.inverse);
+
+        let class = match (self.class, rhs.class) {
+            (TC::Translation, TC::Translation) => TC::Translation,
+            (TC::Translation, TC::Affine) => TC::Affine,
+            (TC::Affine, TC::Translation) => TC::Affine,
+            (TC::Affine, TC::Affine)      => TC::Affine,
+            (_, _) => TC::Projection
+        };
+
+        Proj { transform: t, inverse: i, class: class }
+    }
+}
+
+/// Combines two projective transformations
+/// (matrix multiplication)
+impl<'a, 'b> Mul<&'b Proj> for &'a Proj {
+    type Output = Proj;
+
+    fn mul(self, rhs: &Proj) -> Proj {
+        use TransformationClass as TC;
+        let t = mul3x3(self.transform, rhs.transform);
+        let i = mul3x3(rhs.inverse, self.inverse);
+        let class = match (self.class, rhs.class) {
+            (TC::Translation, TC::Translation) => TC::Translation,
+            (TC::Translation, TC::Affine) => TC::Affine,
+            (TC::Affine, TC::Translation) => TC::Affine,
+            (_, _) => TC::Projection
+        };
+
+        Proj { transform: t, inverse: i, class: class }
+    }
+}
+
+impl Mul<(f32, f32)> for Proj {
+    type Output = (f32, f32);
+
+    fn mul(self, rhs: (f32, f32)) -> (f32, f32) {
+        let (x,y) = rhs;
+        self.warp_p(x, y)
+    }
+}
+
+impl<'a, 'b> Mul<&'b (f32, f32)> for &'a Proj {
+    type Output = (f32, f32);
+
+    fn mul(self, rhs: &(f32, f32)) -> (f32, f32) {
+        let (x,y) = *rhs;
+        self.warp_p(x, y)
+    }
+}
+
+/// Performs projective transformation to an image. Allocates an output image with same
+/// dimensions as the input image. Output pixels outside the input image are set to `default`.
+/// Proj `homography` defines a mapping from coordinates in the `output` image to coordinates of
+/// the `image`.
+pub fn warp_new<P>(homography: &Proj,
+                   image: &Image<P>,
+                   interpolation: Interpolation,
+                   default: P
+                  ) -> Image<P> 
+where
```

Comment:
The formatting of the function parameters and the where clauses don't match the rate of this library.

Reply:
Fixed. That was sloppy, sorry.